### PR TITLE
Val bs fix and distance computation refactor

### PR DESCRIPTION
**Pull request by M-R-Schaefer**
This PR fixes a bug in the data pipeline where the validation batch size was not used and instead the train batch size was used for everything.
Further:
- I tested the max l2cache option and found it to not do anything, hence I removed it.
- TFPipeline was renamed to AtomisticDataset
- AtomisticDataset is no longer initialized with the batch size
- creating a tf.data dataset from dicts was moved to a separate function
- some type hints
- callback initialization was moved to separate submodule
- the calculation of pairwise distances from cartesian positions was moved from the descriptor and empirical corrections to the energy model. This greatly simplifies the call signature of these models/layers and reduces code duplication.

### CODE REVIEW

File: apax/data/input_pipeline.py [hunk 1]
```diff
@@ -9,15 +11,15 @@
 log = logging.getLogger(__name__)
 
 
-def find_largest_system(inputs):
+def find_largest_system(inputs: dict[np.ndarray]) -> tuple[int]:
```
Comment (PythonFZ) [inline apax/data/input_pipeline.py:14]: I think it should be `dict[str, np.ndarray]` ?

File: apax/data/statistics.py [hunk 2]
```diff
@@ -36,7 +34,7 @@ def compute(inputs, labels, shift_options) -> np.ndarray:
         n_atoms_total = np.sum(system_sizes)
 
         mean_energy = ds_energy / n_atoms_total
-        n_species = 119  # max([max(n) for n in numbers]) + 1
+        n_species = 119
```
Comment (PythonFZ) [inline apax/data/statistics.py:37]: I do think an explanation why this number is set to 119 is a good idea.

File: apax/model/gmnn.py [hunk 3]
```diff
@@ -22,6 +25,34 @@
 log = logging.getLogger(__name__)
 
 
+def canonicalize_neighbors(neighbor):
+    if type(neighbor) in [partition.NeighborList, NeighborSpoof]:
```
Comment (PythonFZ) [inline apax/model/gmnn.py:29]: If you are not explicitly excluding subclasses, I'd suggest using `if isinstance(neighbor, (partition.NeighborList, NeighborSpoof))` instead

Comment (PythonFZ) [reply apax/model/gmnn.py:29]: This basically makes it a oneliner:
```py
return neighbor.idx if isinstance(...) else neighbor
```

File: apax/model/gmnn.py [hunk 4]
```diff
@@ -65,27 +101,44 @@ def __call__(
...
             total_energy = total_energy + energy_correction
+
+        # TODO think of nice abstraction for predicting additional properties
         return total_energy
 
 
 class EnergyDerivativeModel(nn.Module):
-    atomistic_model: AtomisticModel = AtomisticModel()
+    # Alternatively, should this be a function transformation?
+    energy_model: EnergyModel = EnergyModel()
```
Comment (PythonFZ) [inline apax/model/gmnn.py:138]: I'm not too familiar with how `nn.Module` are designed to be subclassed, but creating an instance on class level often can lead to issues, therefore I'd suggest moving this to the `__init__` and using `self.energy_model = EnergyModel()` on an instance bases. But I do remember that some of these didn't like an subclassed `__init__` so maybe don't?

File: apax/train/callbacks.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    log.info("Initializing Callbacks")
+
+    dummy_model = tf.keras.Model()
+    callback_dict = {
+        "csv": {
+            "class": CSVLogger,
+            "log_path": model_version_path / "log.csv",
+            "path_arg_name": "filename",
+            "kwargs": {"append": True},
+            "model": dummy_model,
```
Comment (PythonFZ) [inline apax/train/callbacks.py:19]: Could be a datalcass using

```py

@dataclass
class Callback:
    class: object
    log_path: Path
    path_arg_name: str
    kwargs: dict
    model: object

# and
dataclasses.asdict(...)
```

if more loggers are supposed to be added in the future.

File: apax/train/callbacks.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    if "csv" in callback_configs and "tensorboard" in callback_configs:
+        csv_idx, tb_idx = callback_configs.index("csv"), callback_configs.index(
+            "tensorboard"
+        )
+        msg = (
+            "Using both csv and tensorboard callbacks is not supported at the moment."
+            " Only the first of the two will be used."
+        )
+        print("Warning: " + msg)
+        log.warning(msg)
```
Comment (PythonFZ) [inline apax/train/callbacks.py:40]: Why are you using both?

File: apax/train/callbacks.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    callbacks = []
+    for callback_config in callback_configs:
+        callback_info = callback_dict[callback_config]
+
+        path_arg_name = callback_info["path_arg_name"]
+        path = {path_arg_name: callback_info["log_path"]}
+
+        kwargs = callback_info["kwargs"]
+        callback = callback_info["class"](**path, **kwargs)
+        callback.set_model(callback_info["model"])
```
Comment (PythonFZ) [inline apax/train/callbacks.py:55]: If you are using a dataclass, this could be a method of the `Callback` dataclass

File: apax/train/run.py [hunk 8]
```diff
@@ -194,15 +38,26 @@ def setup_logging(log_file, log_level):
     logging.basicConfig(filename=log_file, level=log_levels[log_level])
 
 
+def initialize_directories(model_version_path: Path) -> None:
+    log.info("Initializing directories")
+    os.makedirs(model_version_path, exist_ok=True)
```
Comment (PythonFZ) [inline apax/train/run.py:43]: if the `model_version_path` is truely a `Path` why not use `Path.mkdir(exists_ok=True)`

**Review by PythonFZ — COMMENTED** [review]
I've added a few comments that might or might not be fixed.
In general it looks good, but I can not talk about any of the details - because I'm not familiar enough with the code.

File: apax/model/gmnn.py (hunk 3 above)
Comment (M-R-Schaefer) [reply apax/model/gmnn.py:29]: That's a great suggestion, I would have expected this to give an error but it works.

File: apax/model/gmnn.py (hunk 4 above)
Comment (M-R-Schaefer) [reply apax/model/gmnn.py:138]: yeah there is no overwriting `__init__` for flax and `setup` doesn't take arguments. So as far as I am aware, this is the way to go with flax.

File: apax/train/callbacks.py (hunk 6 above)
Comment (M-R-Schaefer) [reply apax/train/callbacks.py:40]: I am not terribly happy with our logging setup. Log level is set to warn by default and logs are written to a file. I would like to do something like write log.info to a file and log.warning to both stdout and the log file.
Reason being that I think users encountering a warning might not check the log file to see what's going on, but if it's in the terminal, they are going to see it.

Comment (M-R-Schaefer) [reply apax/train/callbacks.py:40]: but I'm not sure what the best way to handle this is

File: apax/train/run.py (hunk 8 above)
Comment (M-R-Schaefer) [reply apax/train/run.py:43]: good point. The function is pretty trivial by now. In the past it used to do much more. I will just replace it with an inline log and Path.mkdir call

File: apax/train/callbacks.py (hunk 7 above)
Comment (M-R-Schaefer) [reply apax/train/callbacks.py:55]: true, I moved this to an issue.

**Pull request merged**